### PR TITLE
chore: merge data into redshift to avoid duplicate data

### PR DESCRIPTION
**Pull request by dengmingtong**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

merge data into redshift to avoid duplicate data

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [ ] add new test cases
- [x] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [x] deploy data modeling
    - [x] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting
  - [ ] streaming ingestion
    - [ ] with Redshift Serverless
    - [ ] with provisioned Redshift

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/analytics/private/sqls/redshift/sp-merge-event-v2.sql [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+DECLARE
+	log_name varchar(50) := 'sp_merge_event_v2';
+BEGIN
+	CREATE temp TABLE event_v2_stage (like {{schema}}.event_v2);
+
+	EXECUTE 'COPY event_v2_stage FROM ''' || manifestFileName || ''' IAM_ROLE ''' || iam_role || ''' STATUPDATE ON FORMAT AS PARQUET SERIALIZETOJSON MANIFEST ACCEPTINVCHARS'; 
+	CALL {{schema}}.{{sp_clickstream_log}}(log_name, 'info', 'Copy ods data into event v2 stage table successfully.');
+
+	MERGE INTO {{schema}}.event_v2
+	USING event_v2_stage AS stage on ({{schema}}.event_v2.event_id = stage.event_id)
```
Comment (zxkane) [inline src/analytics/private/sqls/redshift/sp-merge-event-v2.sql:13]: add `event_timestamp` too

File: src/analytics/private/sqls/redshift/sp-merge-item-v2.sql [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+BEGIN
+
+	CREATE temp TABLE item_v2_stage (like {{schema}}.item_v2);
+
+	EXECUTE 'COPY item_v2_stage FROM ''' || manifestFileName || ''' IAM_ROLE ''' || iam_role || ''' STATUPDATE ON FORMAT AS PARQUET SERIALIZETOJSON MANIFEST ACCEPTINVCHARS'; 
+
+	CALL {{schema}}.{{sp_clickstream_log}}(log_name, 'info', 'Copy ods data into item v2 stage table successfully.');
+
+	MERGE INTO {{schema}}.item_v2
+	USING item_v2_stage AS stage ON ({{schema}}.item_v2.event_id = stage.event_id AND {{schema}}.item_v2.user_pseudo_id = stage.user_pseudo_id AND {{schema}}.item_v2.item_id = stage.item_id)
```
Comment (zxkane) [inline src/analytics/private/sqls/redshift/sp-merge-item-v2.sql:15]: add `event_timestamp`

File: src/analytics/private/sqls/redshift/sp-merge-session.sql [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+	CREATE temp TABLE session_stage (like {{schema}}.session);
+
+	EXECUTE 'COPY session_stage FROM ''' || manifestFileName || ''' IAM_ROLE ''' || iam_role || ''' STATUPDATE ON FORMAT AS PARQUET SERIALIZETOJSON MANIFEST ACCEPTINVCHARS'; 
+
+
+	CALL {{schema}}.{{sp_clickstream_log}}(log_name, 'info', 'Copy ods data into session stage table successfully.');
+
+	MERGE INTO {{schema}}.session
+	USING session_stage AS stage ON ({{schema}}.session.user_pseudo_id = stage.user_pseudo_id AND {{schema}}.session.session_id = stage.session_id)
```
Comment (zxkane) [inline src/analytics/private/sqls/redshift/sp-merge-session.sql:16]: add `event_timestamp`

File: src/analytics/private/sqls/redshift/sp-merge-user-v2.sql [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+  EXECUTE 'COPY user_v2_stage FROM ''' || manifestFileName || ''' IAM_ROLE ''' || iam_role || ''' STATUPDATE ON FORMAT AS PARQUET SERIALIZETOJSON MANIFEST ACCEPTINVCHARS'; 
+	
+	CALL {{schema}}.{{sp_clickstream_log}}(log_name, 'info', 'Copy ods data into user v2 stage table successfully.');
+
+	MERGE INTO {{schema}}.user_v2
+	USING user_v2_stage AS stage ON ({{schema}}.user_v2.user_pseudo_id = stage.user_pseudo_id)
+	WHEN MATCHED THEN
+	UPDATE SET 
+			created_time = CURRENT_TIMESTAMP
```
Comment (zxkane) [inline src/analytics/private/sqls/redshift/sp-merge-user-v2.sql:18]: update all user properties to keep the latest user properties

File: test/analytics/analytics-on-redshift/lambda/load-data/load-manifest-to-redshift.test.ts [hunk 5]
```diff
@@ -208,13 +208,7 @@ describe('Lambda - do loading manifest to Provisioned Redshift via COPY command'
   test('Executed Redshift copy command', async () => {
```
Comment (zxkane) [inline test/analytics/analytics-on-redshift/lambda/load-data/load-manifest-to-redshift.test.ts:208]: update case description too

File: src/analytics/private/sqls/redshift/sp-merge-item-v2.sql [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	MERGE INTO {{schema}}.item_v2
+	USING item_v2_stage AS stage ON (
+		{{schema}}.item_v2.event_id = stage.event_id AND 
+		{{schema}}.item_v2.user_pseudo_id = stage.user_pseudo_id AND 
+		{{schema}}.item_v2.item_id = stage.item_id AND 
+		{{schema}}.item_v2.event_timestamp = stage.event_timestamp
+	)
+	WHEN MATCHED THEN
+	UPDATE SET 
+			created_time = CURRENT_TIMESTAMP
```
Comment (llmin) [inline src/analytics/private/sqls/redshift/sp-merge-item-v2.sql:23]: update all item properties

**Review by zxkane — APPROVED** [review]

**Review by amliuyong — APPROVED** [review]

**Review by llmin — APPROVED** [review]

**Pull request merged**